nemo/collections/tts/losses/radttsloss.py

```diff
@@ -0,0 +1,189 @@
+            if len(log_det_W_list):
+                log_det_W_total += log_det_W_list[i]
+
+    if len(log_det_W_list):
+        log_det_W_total *= n_elements
+
+    z = z * mask
+    prior_NLL = torch.sum(z * z) / (2 * sigma * sigma)
+
+    loss = prior_NLL - log_s_total - log_det_W_total
+
+    denom = n_elements * n_dims
+    loss = loss / denom
+    loss_prior = prior_NLL / denom
+    return loss, loss_prior
+
+
+def compute_regression_loss(x_hat, x, mask, name=False):
+    x = x[:, None] if len(x.shape) == 2 else x  # add channel dim
+    mask = mask[:, None] if len(mask.shape) == 2 else mask  # add channel dim
+    assert len(x.shape) == len(mask.shape)
+
+    x = x * mask
+    x_hat = x_hat * mask
+
+    if name == 'vpred':
+        loss = F.binary_cross_entropy_with_logits(x_hat, x, reduction='sum')
+    else:
+        loss = F.mse_loss(x_hat, x, reduction='sum')
+    loss = loss / mask.sum()
+
+    loss_dict = {"loss_{}".format(name): loss}
+
+    return loss_dict
+
+
+class AttributePredictionLoss(torch.nn.Module):
+    def __init__(self, name, model_config, loss_weight, sigma=1.0):
+        super(AttributePredictionLoss, self).__init__()
+        self.name = name
+        self.sigma = sigma
+        self.model_name = model_config['name']
+        self.loss_weight = loss_weight
+        self.n_group_size = 1
+        if 'n_group_size' in model_config['hparams']:
+            self.n_group_size = model_config['hparams']['n_group_size']
+
+    def forward(self, model_output, lens):
+        mask = get_mask_from_lengths(lens // self.n_group_size)
+        mask = mask[:, None].float()
+        loss_dict = {}
+        if 'z' in model_output:
+            n_elements = lens.sum() // self.n_group_size
+            n_dims = model_output['z'].size(1)
+
+            loss, loss_prior = compute_flow_loss(
+                model_output['z'],
+                model_output['log_det_W_list'],
+                model_output['log_s_list'],
+                n_elements,
+                n_dims,
+                mask,
+                self.sigma,
+            )
+            loss_dict = {
+                "loss_{}".format(self.name): (loss, self.loss_weight),
+                "loss_prior_{}".format(self.name): (loss_prior, 0.0),
+            }
+        elif 'x_hat' in model_output:
+            loss_dict = compute_regression_loss(model_output['x_hat'], model_output['x'], mask, self.name)
+            for k, v in loss_dict.items():
+                loss_dict[k] = (v, self.loss_weight)
+
+        if len(loss_dict) == 0:
+            raise Exception("loss not supported")
+
+        return loss_dict
+
+
+class AttentionBinarizationLoss(torch.nn.Module):
+    def __init__(self):
+        super(AttentionBinarizationLoss, self).__init__()
+
+    def forward(self, hard_attention, soft_attention):
+        log_sum = torch.log(soft_attention[hard_attention == 1]).sum()
+        return -log_sum / hard_attention.sum()
+
+
+class RADTTSLoss(torch.nn.Module):
```

Comment:
I still see a `torch.nn.Module` here when it should be `nemo.core.classes.Loss`

Reply:
Using nemo.core.classes.Loss now